velox/functions/prestosql/Zip.cpp

```diff
@@ -0,0 +1,221 @@
+      const TypePtr& outputType,
+      exec::EvalCtx* context,
+      VectorPtr* result) const override {
+    validateInputTypes(args);
+    const vector_size_t numInputArrays = args.size();
+
+    exec::DecodedArgs decodedArgs(rows, args, context);
+    std::vector<const ArrayVector*> baseVectors(args.size());
+    std::vector<const vector_size_t*> rawSizes(args.size());
+    std::vector<const vector_size_t*> rawOffsets(args.size());
+    std::vector<const vector_size_t*> indices(args.size());
+
+    for (int i = 0; i < numInputArrays; i++) {
+      baseVectors[i] = decodedArgs.at(i)->base()->as<ArrayVector>();
+      rawSizes[i] = baseVectors[i]->rawSizes();
+      rawOffsets[i] = baseVectors[i]->rawOffsets();
+      indices[i] = decodedArgs.at(i)->indices();
+    }
+
+    // Size of elements in result vector.
+    vector_size_t resultElementsSize = 0;
+
+    // Determine what the size of the resultant elements will be so we can
+    // reserve enough space.
+    auto getMaxArraySize = [&](vector_size_t row) -> vector_size_t {
+      vector_size_t maxSize = 0;
+      for (int i = 0; i < numInputArrays; i++) {
+        maxSize = std::max(maxSize, rawSizes[i][indices[i][row]]);
+      }
+      return maxSize;
+    };
+
+    BufferPtr maxArraySizesBuffer =
+        AlignedBuffer::allocate<vector_size_t>(rows.end(), context->pool());
+    auto maxArraySizes = maxArraySizesBuffer->asMutable<vector_size_t>();
+    rows.applyToSelected([&](auto row) {
+      auto maxSize = getMaxArraySize(row);
+      resultElementsSize += maxSize;
+      maxArraySizes[row] = maxSize;
+    });
+
+    // Now create individual result vectors for each input Array vector.
+
+    auto pool = context->pool();
+    std::vector<BufferPtr> resultIndices(numInputArrays);
+    std::vector<BufferPtr> resultNulls(numInputArrays);
+    std::vector<vector_size_t*> rawResultIndices(numInputArrays);
+    std::vector<uint64_t*> rawResultNulls(numInputArrays);
+
+    for (int i = 0; i < numInputArrays; i++) {
+      resultIndices[i] = allocateIndices(resultElementsSize, pool);
+      resultNulls[i] = AlignedBuffer::allocate<bool>(
+          resultElementsSize, pool, bits::kNotNull);
+      rawResultIndices[i] = resultIndices[i]->asMutable<vector_size_t>();
+      rawResultNulls[i] = resultNulls[i]->asMutable<uint64_t>();
+    }
+
+    // Final Row Vector
+    const auto finalArraySize = rows.end();
+    BufferPtr finalArrayIndices = allocateIndices(finalArraySize, pool);
+    BufferPtr finalArrayNulls =
+        AlignedBuffer::allocate<bool>(finalArraySize, pool, bits::kNotNull);
+    BufferPtr finalArrayOffsets = allocateOffsets(finalArraySize, pool);
+    BufferPtr finalArraySizes = allocateSizes(finalArraySize, pool);
+
+    auto rawFinalArrayOffsets = finalArrayOffsets->asMutable<vector_size_t>();
+    auto rawFinalArraySizes = finalArraySizes->asMutable<vector_size_t>();
+
+    // Create right offsets/indexes for the individual and final result arrays.
+    int rowIndex = 0;
```

Comment:
nit: would it be clearer if this variable was renamed to nestedRow or elementRow?